MIDI Audio rendering panics when playing chords

If more than one note is played at once, you'll get this error:
```
panic: runtime error: index out of range [1206] with length 1188            

goroutine 653 [running]:
github.com/sinshu/go-meltysynth/meltysynth.(*combFilter).process(0xc0040e7830, {0xc0040ee200, 0x40, 0x3ed2b8f6002c3a8e?}, {0xc0040ee300, 0x40, 0xc002695d68?})
        /github.com/sinshu/go-meltysynth@v0.0.0-20230125141251-0af16dc927d3/meltysynth/reverb.go:266 +0x1f0
github.com/sinshu/go-meltysynth/meltysynth.(*reverb).process(0xc0040d9c20, {0xc0040ee200, 0x40, 0x40}, {0xc0040ee300?, 0x40, 0x40}, {0xc0040ee400, 0x40, 0x40})
        /github.com/sinshu/go-meltysynth@v0.0.0-20230125141251-0af16dc927d3/meltysynth/reverb.go:146 +0xbc
github.com/sinshu/go-meltysynth/meltysynth.(*Synthesizer).renderBlock(0xc004076b00)
        /github.com/sinshu/go-meltysynth@v0.0.0-20230125141251-0af16dc927d3/meltysynth/synthesizer.go:394 +0x55b
github.com/sinshu/go-meltysynth/meltysynth.(*Synthesizer).Render(0xc004076b00, {0xc00256e000, 0x15888, 0xc002695e98?}, {0xc0025dc000, 0x15888, 0x10034cba0?})
        /github.com/sinshu/go-meltysynth@v0.0.0-20230125141251-0af16dc927d3/meltysynth/synthesizer.go:326 +0x71
github.com/rapidmidiex/rmxtui/midi.Player.Play({0xc00011b290?, 0xc004076b00?, 0xc00405d980?}, {0xc002695f84?, 0x18?, 0xc000100900?}, 0xc0040cfa00?)
        /github.com/rapidmidiex/rmx-tui-client/midi/midi.go:83 +0x8b
github.com/rapidmidiex/rmxtui/jamui.(*model).playMIDI.func1()
        /github.com/rapidmidiex/rmx-tui-client/jamui/jam.go:431 +0x130
```

https://github.com/sinshu/go-meltysynth/blob/bd68d4de740f37e996e5e5b84e80a72223626f71/meltysynth/reverb.go#L337
https://github.com/sinshu/go-meltysynth/blob/bd68d4de740f37e996e5e5b84e80a72223626f71/meltysynth/reverb.go#L146
https://github.com/sinshu/go-meltysynth/blob/bd68d4de740f37e996e5e5b84e80a72223626f71/meltysynth/synthesizer.go#L394
https://github.com/sinshu/go-meltysynth/blob/bd68d4de740f37e996e5e5b84e80a72223626f71/meltysynth/synthesizer.go#L326
`p.synth.Render(streamer.left, streamer.right)`
https://github.com/Rapidmidiex/rmxtui/blob/feat/midi-to-audio/midi/midi.go#L83

Needs investigation

--- a/jamui/jam.go
+++ b/jamui/jam.go
@@ -5,6 +5,7 @@ import (
 	"log"
 	"os"
 	"strings"
+	"sync"
 	"time"
 
 	"github.com/charmbracelet/bubbles/key"
@@ -89,8 +90,10 @@ type (
 		pianoNotes vpiano.Notes
 		// Currently active piano keys
 		activeKeys map[string]struct{}
-		// Websocket connection for current Jam Session
-		Socket *websocket.Conn
+
+		// Websocket client
+		wsClient *wsClient
+
 		// Jam Session ID
 		ID string
 		// Chat container
@@ -107,12 +110,23 @@ type (
 		userName  string
 		userID    uuid.UUID
 
-		curMidiMsg wsmsg.MIDIMsg
-		midiPlayer midi.Synth
-		mixer      *beep.Mixer
-		sampleRate beep.SampleRate
-		noteKeyMap vpiano.NoteKeyMap
-		log        *log.Logger
+		curMidiMsg  wsmsg.MIDIMsg
+		midiPlayer  midi.Synth
+		audioPlayer *audioPlayer
+		sampleRate  beep.SampleRate
+		noteKeyMap  vpiano.NoteKeyMap
+		log         *log.Logger
+	}
+
+	wsClient struct {
+		mu sync.Mutex
+		// Websocket connection for current Jam Session
+		conn *websocket.Conn
+	}
+
+	audioPlayer struct {
+		mu    sync.Mutex
+		mixer *beep.Mixer
 	}
 )
 
@@ -147,16 +161,16 @@ func New() (model, error) {
 		// If more focus states are added, update number of available states
 		availableFocusStates: 2,
 
-		rtTimer:    rtt.NewTimer(),
-		pingStats:  rtt.NewStats(),
-		noteKeyMap: pianoNotes.ToBindingMap(),
-		midiPlayer: midiPlayer,
-		mixer:      &beep.Mixer{},
-		sampleRate: sr,
-		log:        log.Default(),
+		rtTimer:     rtt.NewTimer(),
+		pingStats:   rtt.NewStats(),
+		noteKeyMap:  pianoNotes.ToBindingMap(),
+		midiPlayer:  midiPlayer,
+		audioPlayer: &audioPlayer{mixer: &beep.Mixer{}},
+		sampleRate:  sr,
+		log:         log.Default(),
 	}
 
-	speaker.Play(m.mixer)
+	speaker.Play(m.audioPlayer.mixer)
 	return m, nil
 }
 
@@ -199,7 +213,7 @@ func (m model) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
 
 	// Entered the Jam Session
 	case ConnectedMsg:
-		m.Socket = msg.WS
+		m.wsClient = &wsClient{conn: msg.WS}
 		m.ID = msg.JamID
 		cmds = append(cmds, m.listenSocket())
 
@@ -270,11 +284,11 @@ func (m model) View() string {
 // LeaveRoom disconnects from the room and sends a LeaveRoom message.
 func (m model) leaveRoom() tea.Cmd {
 	return func() tea.Msg {
-		if m.Socket == nil {
+		if m.wsClient.conn == nil {
 			return LeaveRoomMsg{}
 		}
 		// Send websocket close message
-		err := m.Socket.WriteControl(
+		err := m.wsClient.conn.WriteControl(
 			websocket.CloseMessage,
 			nil,
 			time.Now().Add(time.Second*10),
@@ -291,7 +305,7 @@ func (m model) listenSocket() tea.Cmd {
 	// https://github.com/charmbracelet/bubbletea/issues/25#issuecomment-732339162
 	return func() tea.Msg {
 		var message wsmsg.Envelope
-		err := m.Socket.ReadJSON(&message)
+		err := m.wsClient.readMsg(&message)
 		if err != nil {
 			if websocket.IsUnexpectedCloseError(err, websocket.CloseGoingAway, websocket.CloseAbnormalClosure) {
 				return rmxerr.ErrMsg{Err: fmt.Errorf("readJSON: unexpected close: %w", err)}
@@ -358,7 +372,7 @@ func (m model) sendTextMessage(body string) tea.Cmd {
 
 		// Curious to see how long WriteJSON takes
 		preSendTime := time.Now()
-		err = m.Socket.WriteJSON(envelope)
+		err = m.wsClient.conn.WriteJSON(envelope)
 		if err != nil {
 			return rmxerr.ErrMsg{Err: fmt.Errorf("writeJSON: %w", err)}
 		}
@@ -391,8 +405,7 @@ func (m model) sendMIDIMessage(keyPressed string) tea.Cmd {
 		if err := envelope.SetPayload(msg); err != nil {
 			return rmxerr.ErrMsg{Err: fmt.Errorf("marshal: %w", err)}
 		}
-
-		if err := m.Socket.WriteJSON(envelope); err != nil {
+		if err := m.wsClient.writeMsg(envelope); err != nil {
 			return rmxerr.ErrMsg{Err: fmt.Errorf("writeJSON: %w", err)}
 		}
 		return sentMsg{
@@ -410,12 +423,15 @@ func (m *model) playMIDI(note wsmsg.MIDIMsg) tea.Cmd {
 		// TODO: Maybe control duration with some other key
 		duration := time.Second * 2
 		s := midi.NewMIDIStreamer(duration)
+
 		// Render MIDI note to audio streamer buffer
-		m.midiPlayer.Render(note, s)
+		if err := m.midiPlayer.Render(note, s); err != nil {
+			return rmxerr.ErrMsg{Err: err}
+		}
 
 		// Take n seconds worth of samples @ 44.1khz from the audio streamer and
 		// add it to the main speaker mix.
-		m.mixer.Add(beep.Take(m.sampleRate.N(duration), s))
+		m.audioPlayer.addToMix(beep.Take(m.sampleRate.N(duration), s))
 
 		return nil
 	}
@@ -433,4 +449,20 @@ func (m model) renderPiano() string {
 		)
 	}
 	return lipgloss.JoinHorizontal(lipgloss.Top, pianoKeys...)
+}
+
+func (c *wsClient) readMsg(out *wsmsg.Envelope) error {
+	return c.conn.ReadJSON(out)
+}
+
+func (c *wsClient) writeMsg(envelope wsmsg.Envelope) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return c.conn.WriteJSON(envelope)
+}
+
+func (a *audioPlayer) addToMix(s beep.Streamer) {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	a.mixer.Add(s)
 }

--- a/midi/midi.go
+++ b/midi/midi.go
@@ -33,6 +33,7 @@ type (
 	}
 
 	MidiStreamer struct {
+		err   error
 		pos   int
 		left  []float32
 		right []float32
@@ -104,8 +105,11 @@ func (ms *MidiStreamer) Stream(samples [][2]float64) (n int, ok bool) {
 	left := make([]float32, len(samples))
 	right := make([]float32, len(samples))
 
-	ms.Read(left, right)
-
+	_, err := ms.Read(left, right)
+	if err != nil {
+		ms.err = err
+		return
+	}
 	for i := range samples {
 		samples[i][0] = float64(left[i])
 		samples[i][1] = float64(right[i])
@@ -134,7 +138,7 @@ func (ms *MidiStreamer) Seek(p int) error {
 }
 
 func (ms MidiStreamer) Err() error {
-	return nil
+	return ms.err
 }
 
 // Read reads from the MIDIStreamer's left/right buffers at the current Pos and writes the contents to the out []float32 buffers.

--- a/midi/midi_test.go
+++ b/midi/midi_test.go
@@ -18,7 +18,9 @@ func TestMIDItoAudio(t *testing.T) {
 	// Lower -> more CPU, faster response
 	bufLen := sr.N(time.Millisecond * 20)
 	noteDuration := time.Second * 5
-	speaker.Init(sr, bufLen)
+	err := speaker.Init(sr, bufLen)
+	require.NoError(t, err)
+
 	synth, err := midi.NewSynth(midi.NewSynthOpts{
 		SoundFontName: midi.GeneralUser,
 	})
@@ -32,7 +34,8 @@ func TestMIDItoAudio(t *testing.T) {
 		}
 
 		streamer := midi.NewMIDIStreamer(noteDuration)
-		synth.Render(msg, streamer)
+		err := synth.Render(msg, streamer)
+		require.NoError(t, err)
 
 		done := make(chan bool)
 
@@ -50,25 +53,28 @@ func TestMIDItoAudio(t *testing.T) {
 
 	t.Run("plays multiple notes at once", func(t *testing.T) {
 		streamer1 := midi.NewMIDIStreamer(noteDuration)
-		synth.Render(wsmsg.MIDIMsg{
+		err := synth.Render(wsmsg.MIDIMsg{
 			State:    wsmsg.NOTE_ON,
 			Number:   70, // Bb4
 			Velocity: 127,
 		}, streamer1)
+		require.NoError(t, err)
 
 		streamer2 := midi.NewMIDIStreamer(noteDuration)
-		synth.Render(wsmsg.MIDIMsg{
+		err = synth.Render(wsmsg.MIDIMsg{
 			State:    wsmsg.NOTE_ON,
 			Number:   67, // G4
 			Velocity: 127,
 		}, streamer2)
+		require.NoError(t, err)
 
 		streamer3 := midi.NewMIDIStreamer(noteDuration)
-		synth.Render(wsmsg.MIDIMsg{
+		err = synth.Render(wsmsg.MIDIMsg{
 			State:    wsmsg.NOTE_ON,
 			Number:   76, // E5
 			Velocity: 108,
 		}, streamer3)
+		require.NoError(t, err)
 
 		mixer := beep.Mixer{}
 		mixer.Add(
@@ -93,24 +99,27 @@ func TestMIDItoAudio(t *testing.T) {
 
 	t.Run("plays chords", func(t *testing.T) {
 		streamer1 := midi.NewMIDIStreamer(noteDuration)
-		synth.Render(wsmsg.MIDIMsg{
+		err := synth.Render(wsmsg.MIDIMsg{
 			State:    wsmsg.NOTE_ON,
 			Number:   40,
 			Velocity: 127,
 		}, streamer1)
+		require.NoError(t, err)
 
 		streamer2 := midi.NewMIDIStreamer(noteDuration)
-		synth.Render(wsmsg.MIDIMsg{
+		err = synth.Render(wsmsg.MIDIMsg{
 			State:    wsmsg.NOTE_ON,
 			Number:   42,
 			Velocity: 108,
 		}, streamer2)
+		require.NoError(t, err)
 
 		mixer := beep.Mixer{}
 		mixer.Add(
 			beep.Take(sr.N(noteDuration), streamer1),
 			beep.Take(sr.N(noteDuration), streamer2),
 		)
+		require.NoError(t, mixer.Err())
 
 		// Only need to call Play once.
 		// The mixer will play silence if the streamers are drained.

